File: ch10/ch10_ws/src/learn_lifecyclenode_py/learn_lifecyclenode_py/learn_lifecyclenode.py
import rclpy
from rclpy.lifecycle import LifecycleNode, TransitionCallbackReturn

class LearnLifeCycleNode(LifecycleNode):
    def __init__(self):
        super().__init__('lifecyclenode')
        self.timer_period = 0
        self.timer_ = None
        self.get_logger().info(f'{self.get_name()}: Created')

    def timer_callback(self):
        self.get_logger().info('Timer printing in progress...')

    def on_configure(self, state):
        self.timer_period = 1.0  # Set timer period
        self.get_logger().info('on_configure(): Configure timer_period')
        return TransitionCallbackReturn.SUCCESS

    def on_activate(self, state):
        self.timer_ = self.create_timer(self.timer_period, self.timer_callback)
        self.get_logger().info('on_activate(): Handle activate command, create timer')
        return TransitionCallbackReturn.SUCCESS

    def on_deactivate(self, state):
        self.destroy_timer(self.timer_) # Destroy timer
        self.get_logger().info('on_deactivate(): Handle deactivate command, stop timer')
        return TransitionCallbackReturn.SUCCESS

    def on_cleanup(self, state):
        self.timer_ = None
        self.timer_period = 0
        self.get_logger().info('on_cleanup(): Handle cleanup command')
        return TransitionCallbackReturn.SUCCESS

    def on_shutdown(self, state):
        # Destroy timer if not already destroyed
        if self.timer_: self.destroy_timer(self.timer_)
        self.get_logger().info('on_shutdown(): Handle shutdown command')
        return TransitionCallbackReturn.SUCCESS

    def on_error(self, state):
        # Directly call parent class handler
        return super().on_error(state)

def main():
    rclpy.init()
    node = LearnLifeCycleNode()
    rclpy.spin(node)
    rclpy.shutdown()

File: ch10/ch10_ws/src/learn_lifecyclenode_cpp/src/learn_lifecyclenode.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_lifecycle/lifecycle_node.hpp"

using CallbackReturn =
    rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn;

class LearnLifeCycleNode : public rclcpp_lifecycle::LifecycleNode {
public:
  LearnLifeCycleNode()
      : rclcpp_lifecycle::LifecycleNode("lifecyclenode") {
    timer_period_ = 1.0;
    timer_ = nullptr;
    RCLCPP_INFO(get_logger(), "%s: Created", get_name());
  }

  CallbackReturn on_configure(const rclcpp_lifecycle::State &state) override {
    (void)state;
    timer_period_ = 1.0;
    RCLCPP_INFO(get_logger(), "on_configure(): Configure timer_period");
    return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::
        CallbackReturn::SUCCESS;
  }

  CallbackReturn on_activate(const rclcpp_lifecycle::State &state) override {
    (void)state;
    timer_ = create_wall_timer(
        std::chrono::seconds(static_cast<int>(timer_period_)),
        [this]() { RCLCPP_INFO(get_logger(), "Timer printing in progress..."); });
    RCLCPP_INFO(get_logger(), "on_activate(): Handle activate command, create timer");
    return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::
        CallbackReturn::SUCCESS;
  }

  CallbackReturn on_deactivate(const rclcpp_lifecycle::State &state) override {
    (void)state;
    timer_.reset();
    RCLCPP_INFO(get_logger(), "on_deactivate(): Handle deactivate command, stop timer");
    return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::
        CallbackReturn::SUCCESS;
  }

  CallbackReturn on_shutdown(const rclcpp_lifecycle::State &state) override {
    (void)state;
    timer_.reset();
    RCLCPP_INFO(get_logger(), "on_shutdown(): Handle shutdown command");
    return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::
        CallbackReturn::SUCCESS;
  }

private:
  rclcpp::TimerBase::SharedPtr timer_;
  double timer_period_;
};

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<LearnLifeCycleNode>();
  rclcpp::spin(node->get_node_base_interface());
  rclcpp::shutdown();
  return 0;
}